### PR TITLE
Make constants easier configurable

### LINKED ISSUE #158: More dynamically configurable options
We have lots of constants inside the application but most of them aren't accessible through `config.conf`. We might want to figure out a more dynamic approach to this rather than adding a line to `read-conf.py` for each new constant.

### PR DESCRIPTION
**Pull request by Androbin**
Fixes #158 

The script `tools/read-conf.py` now lists and describes all constants in `config.conf` as well as the new `custom.conf` which are also extracted from various JS files.
This approach allows for more dynamically configurable options and is easily adaptable.

### CODE REVIEW

**Comment by shardulc** [comment]
@Androbin I think what @sushain97 meant in #158 was that configuration options like all the constants [here](https://github.com/goavki/apertium-html-tools/blob/master/assets/js/translator.js#L10) should be somehow dynamically put into a configuration file with the appropriate sections. That way, all constants which can be customized can be customized through a single file, and constants can be easily accessed through any part of html-tools. (@sushain97 Correct me if I'm wrong!)

Your changes make it so that the _existing_ options are easy to change and add manually, and the code is definitely more maintainable, but they do not solve the problem above. Also, they [break the interface entirely](https://52-28576409-gh.circle-artifacts.com/0/home/circleci/project/build/index.eng.html?dir=ukr-rus#translation) so you need to test your changes.

(P.S. Don't squash your commits on the PR, we'll take care of that when we merge! :smiley:)

**Comment by sushain97** [comment]
Yes, *exactly* what @shardulc said :) Read my mind.

**Comment by sushain97** [comment]
For tasks like this one where the issue doesn't exactly spell out a solution, providing a straw man proposal before making the changes is generally a good idea.

**Comment by Androbin** [comment]
@shardulc I have verified that all options get set correctly. What have I missed?
@sushain97 What about this? I could port the proposed python code to javascript and import it from the files that need it.

**Comment by sushain97** [comment]
> @shardulc I have verified that all options get set correctly. What have I missed?

Check whether there are any JS errors and if the built JS looks the same as it used to.

> @sushain97 What about this? I could port the proposed python code to javascript and import it from the files that need it.

Adding a dependency on Node.js or another JS runner isn't acceptable unfortunately.

**Comment by Androbin** [comment]
> Check whether there are any JS errors and if the built JS looks the same as it used to.

But I have not touched any JavaScript code yet.
> Adding a dependency on Node.js or another JS runner isn't acceptable unfortunately.

Don't get me wrong, @shardulc said 
> [...] constants can be easily accessed through any part of html-tools.

which I read as "one central config file" (conf-fields.txt) and language-specific code to read from that file.

**Comment by sushain97** [comment]
> But I have not touched any JavaScript code yet.

The Python script is used in the generation of JS. Therefore, your changes can easily result in changes in the JS.

> language-specific code to read from that file

This might work provided that the config file is still in a standard format (e.g. YML or INI).

**Comment by Androbin** [comment]
I have just realized that a port to JS might not be needed. If I got this right, [`tools/read-conf.py > print_js`](https://github.com/goavki/apertium-html-tools/pull/216/files#diff-bbd3716c3cae20b9a539236beb8f406eR88) injects a [config global](https://github.com/goavki/apertium-html-tools/blob/master/assets/js/translator.js#L24) into JS.
If I move [all the constants](https://github.com/goavki/apertium-html-tools/blob/master/assets/js/translator.js#L10) into the [config file](https://github.com/Androbin/apertium-html-tools/blob/537e9a648e37df1de07ddbe104616a5ae3051050/tools/conf-fields.txt), then JS should be able to access them.

**Comment by sushain97** [comment]
> If I move all the constants into the config file, then JS should be able to access them.

I don't see this as ideal. There are lots of these and they change often. e.g. if a code refactor renames one of them, it should not require everyone changing their configuration file. I'm fine with everyone needing to change their configuration file if they want their custom value to still work but not requiring them to for the interface to function with defaults. Either way, all the constants shouldn't all be declared in multiple places. Perhaps the ones that are overridable can have a comment in the JS indicating as much.

**Comment by shardulc** [comment]
I think a good solution would be to have two config files: one for 'options' like enabling modes etc., which we already have, and one for 'customizations' of all the constants that are currently within the JS files, which is what you would need to create. Both these files would be read by the same config parser and injected as the same global config.

**Comment by Androbin** [comment]
@shardulc Where would you prefer the config files to be located?

**Comment by shardulc** [comment]
`config.conf` (or rather `config.conf.example`) already exists in the top directory, so I'd say you add the other config file in the top directory too. Try to name it something which clearly marks it as being different from `config.conf`.

**Comment by Androbin** [comment]
Shall I de-duplicate [`tools/conf-fields.txt`](https://github.com/Androbin/apertium-html-tools/blob/537e9a648e37df1de07ddbe104616a5ae3051050/tools/conf-fields.txt) and `config.conf` then?

**Comment by shardulc** [comment]
Yes. Personally I think the type of each constant can be in a dict in `tools/read-conf.py`, although @sushain97 might have a better idea.

**Comment by Androbin** [comment]
@shardulc But the dict could only hold 'options' but not 'customizations', right? What about comments behind each entry?

**Comment by shardulc** [comment]
Why not both? And what comments? (By the way, I would prefer to have this discussion on [IRC](http://wiki.apertium.org/wiki/IRC).)

**Comment by Androbin** [comment]
I have not yet moved the new constants into `config-custom.conf` but other than that, it should work now.

**Comment by Androbin** [comment]
@sushain97 Do you approve the proposal so far?

**Comment by shardulc** [comment]
@Androbin I was thinking that we would not have `config-custom.conf.example` at all, instead directly putting `config-custom.conf` under version control (unlike `config.conf`). Also, the build is still not successful, see the second-last step [here](https://circleci.com/gh/goavki/apertium-html-tools/55#artifacts/containers/0) (I don't know why it's showing up as successful...).

**Comment by Androbin** [comment]
@shardulc Alright, I have figured it out. There is a lower/upper case issue with `configparser`. Gimme a moment to fix this.

**Comment by Androbin** [comment]
Also, some values were not set, but should be set to `None` which didn't work by default.

**Comment by shardulc** [comment]
@Androbin A successful build will produce `index.html` as an artifact in CircleCI (by clicking on the green arrow above), which isn't happening right now. (Or you could run `make` on your computer and make sure it's created.)

**Comment by Androbin** [comment]
@shardulc Just needed to fix a reference.

**Comment by Androbin** [comment]
Many constants were previously overwritten with layers of defaults.
I needed a moment to figure out which one should get set respectively.

**Comment by Androbin** [comment]
I ran a quick `diff` over `master/build` and `gci/build` and apart from the non-deterministic ordering of a few lines, they seem to match up.

**Comment by Androbin** [comment]
I have moved all constants all JS files including imports/exports to the config file and tested everything.
@sushain97 I hope this is to your liking.

**Comment by Androbin** [comment]
@shardulc Would you prefer `config-custom.conf` or `custom.conf` after all?
In addition to `string`, `int`, `bool` and `list` (of `string`), I will add support for `list` of `int`.

File: README.md [hunk 1]
```diff
@@ -24,7 +24,7 @@ Setup
 -------
 
 1. Copy `config.conf.example` to `config.conf` and edit it.
-1. Then type `make`.
+2. Then type `make`.
```
Comment (sushain97) [inline README.md]: Is there a reason you changed this?

File: assets/js/analyzer.js [hunk 2]
```diff
@@ -24,7 +23,7 @@ if(modeEnabled('analyzation')) {
         });
 
         $('#morphAnalyzerInput').keydown(function (e) {
-            if(e.keyCode === ENTER_KEY_CODE && !e.shiftKey) {
+            if(e.keyCode === config.ENTER_KEY_CODE && !e.shiftKey) {
```
Comment (sushain97) [inline assets/js/analyzer.js]: This shouldn't be configurable...

File: assets/js/persistence.js [hunk 3]
```diff
@@ -3,14 +3,12 @@
     populateSecondaryGeneratorList, isSubset, handleNewCurrentLang */
 /* global srcLangs:true, dstLangs:true, recentSrcLangs: true, recentDstLangs:true, curSrcLang:true, curDstLang:true, locale:true */
 
-var URL_PARAM_Q_LIMIT = 1300,
```
Comment (sushain97) [inline assets/js/persistence.js]: This shouldn't be configurable either. The things that are technical limitations and don't make sense to configure, shouldn't be configurable.

File: assets/js/translator.js [hunk 4]
```diff
@@ -7,24 +7,10 @@ var recentSrcLangs = [], recentDstLangs = [];
...
-    TRANSLATION_LIST_WIDTH = 650,
-    TRANSLATION_LIST_ROWS = 8,
-    TRANSLATION_LIST_COLUMNS = 4,
-    TRANSLATION_LISTS_BUFFER = 50;
-
-var INSTANT_TRANSLATION_URL_DELAY = 500,
-    INSTANT_TRANSLATION_PUNCTUATION_DELAY = 1000,
-    INSTANT_TRANSLATION_DELAY = 3000;
-
-var PUNCTUATION_KEY_CODES = [46, 33, 58, 63, 47, 45, 190, 171, 49]; // eslint-disable-line no-magic-numbers
```
Comment (sushain97) [inline assets/js/translator.js]: Should not be configurable.

File: assets/js/util.js [hunk 5]
```diff
@@ -1,22 +1,7 @@
 /* @flow */
 /* exported sendEvent, modeEnabled, filterLangList, getURLParam, onlyUnique, isSubset, safeRetrieve, callApy, apyRequestTimeout, isURL */
-/* exported SPACE_KEY_CODE, ENTER_KEY_CODE, HTTP_OK_CODE, HTTP_BAD_REQUEST_CODE, XHR_LOADING, XHR_DONE */
 /* global config, persistChoices, iso639Codes, iso639CodesInverse, populateTranslationList, showTranslateWebpageInterface */
 
-var SPACE_KEY_CODE = 32, ENTER_KEY_CODE = 13,
```
Comment (sushain97) [inline assets/js/util.js]: None of these should be configurable. They can't be different in any sane world.

File: custom.conf [hunk 6]
```diff
@@ -0,0 +1,35 @@
+[PERSISTENCE]
+URL_PARAM_Q_LIMIT = 1300
+DEFAULT_EXPIRY_HOURS = 24
+
+[TRANSLATOR]
```
Comment (sushain97) [inline custom.conf]: `'TRANSLATION'` for consistency

File: tools/conf-dtypes.txt [hunk 7]
```diff
@@ -0,0 +1,63 @@
+[APY]
```
Comment (sushain97) [inline tools/conf-dtypes.txt]: While this file is OK, I'd much rather we not have it at all. It seems to be largely possible to automatically differentiate between these types.

File: tools/read-conf.py [hunk 8]
```diff
@@ -7,57 +7,82 @@
...
+            parts = line.split('|')
+            parts = [part.strip() for part in parts]
 
-        'SHOW_NAVBAR'                    : conf_APY.getboolean('SHOW_NAVBAR', fallback=False),
+            if len(parts) == 1:
+                section = parts[0][1:-1]
+                dtypes[section] = {}
+            elif len(parts) == 2:
+                key = parts[0]
+                dtype = parts[1]
```
Comment (sushain97) [inline tools/read-conf.py]: `key, dtype = parts`?

File: README.md (hunk 1 above)
Comment (Androbin) [reply README.md]: See commit dfe098bc8d54dcf252a4fc742e3bfc3f659faaa1 but I removed the other line in commit 7c9f41a43e14b5ca6e9f619b2a6c190ff5c7ac19

File: assets/js/persistence.js (hunk 3 above)
Comment (Androbin) [reply assets/js/persistence.js]: But this limitation is not technical, it's practical.

File: assets/js/analyzer.js (hunk 2 above)
Comment (Androbin) [reply assets/js/analyzer.js]: What constants shall I move back to JS?

File: custom.conf (hunk 6 above)
Comment (Androbin) [reply custom.conf]: I named it after `translator.js`

File: README.md (hunk 1 above)
Comment (sushain97) [reply README.md]: OK, let's revert it then?

File: tools/conf-dtypes.txt (hunk 7 above)
Comment (Androbin) [reply tools/conf-dtypes.txt]: But there is a chance that a number is meant to be a string. Also, a flawky config file should not mis-determine the dtype of the value.

File: assets/js/persistence.js (hunk 3 above)
Comment (sushain97) [reply assets/js/persistence.js]: No, it's technical. Some browsers have limitations on maximum URL length.

File: tools/read-conf.py (hunk 8 above)
Comment (Androbin) [reply tools/read-conf.py]: Sure

File: assets/js/analyzer.js (hunk 2 above)
Comment (sushain97) [reply assets/js/analyzer.js]: I think you can figure out the correct set.

File: tools/conf-dtypes.txt (hunk 7 above)
Comment (sushain97) [reply tools/conf-dtypes.txt]: Where should a number be interpreted as a string? That just sounds like a misleading configuration variable that should be replaced with an enum or something.

File: README.md (hunk 1 above)
Comment (Androbin) [reply README.md]: Why, shouldn't it be 1. and 2. instead of 1. and 1. ?

File: assets/js/persistence.js (hunk 3 above)
Comment (Androbin) [reply assets/js/persistence.js]: Yeah, but there is no definite maximum. I'll revert this anyway.

**Comment by sushain97** [comment]
The latter never requires changing irrelevant lines for inserting an item,
i.e. smaller diffs.

On Dec 3, 2017 3:09 PM, "Robin Richtsfeld" <notifications@github.com> wrote:

*@Androbin* commented on this pull request.
------------------------------

In README.md
<https://github.com/goavki/apertium-html-tools/pull/216#discussion_r154535110>
:

> @@ -24,7 +24,7 @@ Setup
 -------

 1. Copy `config.conf.example` to `config.conf` and edit it.
-1. Then type `make`.
+2. Then type `make`.

Why, shouldn't it be 1. and 2. instead of 1. and 1. ?

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/goavki/apertium-html-tools/pull/216#discussion_r154535110>,
or mute the thread

File: tools/conf-dtypes.txt (hunk 7 above)
Comment (Androbin) [reply tools/conf-dtypes.txt]: The line  `CONSTANT_NAME = 42` would be "automatically differentiated" to `int` even if `CONSTANT_NAME` is a `string` and should be `"42"`

Comment (sushain97) [reply tools/conf-dtypes.txt]: I don't deny that it's theoretically possible. Rephrased: Where should a number be interpreted as a string [in a reasonable setting with our set of options or a reasonable extension]?

Comment (sushain97) [reply tools/conf-dtypes.txt]: The balance to be stricken here is between being strict and type safe or being more maintainable and having less repetition. Unless there's a good argument for the former, the latter is preferable.

Comment (Androbin) [reply tools/conf-dtypes.txt]: @shardulc initially wanted a dict in `read-conf.py` but I suggested to have a separate file for that.

Comment (sushain97) [reply tools/conf-dtypes.txt]: I prefer a dict in `read-conf.py` over this as well and over both I prefer not having a list at all. The example configuration files should serve as documentation for the options. If one is unclear, a comment would be allay the concern.

Comment (sushain97) [reply tools/conf-dtypes.txt]: The separate file also introduces a new syntax which is not ideal.

Comment (sushain97) [reply tools/conf-dtypes.txt]: Consider for example how Django does their `settings.py`. There's no exhaustive list of options.

File: README.md (hunk 1 above)
Comment (sushain97) [reply README.md]: The latter never requires changing irrelevant lines for inserting an item,
i.e. smaller diffs.

**Comment by Androbin** [comment]
> It seems to be largely possible to automatically differentiate between these types.

I'll try to match the types using the following regular expressions (with leading and trailing space already trimmed):

Pretty straightforward:
`bool` -> `True|False`
`int` -> `[0-9]+`
`int[]` -> `[0-9]+([, ]+[0-9]+)+`

But what about this?
`string` -> `.+`
`string[]` -> `.+([, ]+.+)+`

**Comment by sushain97** [comment]
Does it make sense for string to be whatever doesn't match any of the above?

**Comment by shardulc** [comment]
Here's where the conflicts come in: booleans have to accomodate 'True', 'Yes', 'On', and '1' as per `configparser`, but all of these can be strings and one could be an integer... plus we wouldn't want just integers, we'd want to detect all numbers, so it's a little more complex than the regex @Androbin suggested.

**Comment by Androbin** [comment]
@shardulc And what about strings with commas inside? I think this goes a bit out of hand here.

**Comment by shardulc** [comment]
I agree with @Androbin. @sushain97, I think it is best to explicitly say what type each config option has. But I would definitely prefer this to be a dictionary inside `read-conf.py` because like @sushain97 said, the extra file is just another file format which we need to document.

**Comment by sushain97** [comment]
Hmm, OK. This does seem too tricky to be worth it. In that case, a dictionary at the top of `read-conf` probably makes the most sense where the keys are config item names and the values are parsing functions.

**Comment by Androbin** [comment]
@shardulc @sushain97 Instead of a dict, what about a list of names for each type?

**Comment by shardulc** [comment]
That's just confusing. It makes sense to have them in the same order in the config files and in `read-conf.py`.

**Comment by sushain97** [comment]
+ diffs aren't as clear

**Comment by Androbin** [comment]
Okay, gimme a moment.

**Comment by Androbin** [comment]
BTW. Would you prefer a dict to `"string"` or `get_string`?

**Comment by sushain97** [comment]
> BTW. Would you prefer a dict to "string" or get_string?

The function itself, i.e. the latter. This prevents having to case on the string later.

File: config.conf.example [hunk 9]
```diff
@@ -2,12 +2,13 @@
 HTML_URL = https://www.apertium.org
 APY_URL = https://www.apertium.org/apy
 
-#SUBTITLE =
-#SUBTITLE_COLOR =
+SUBTITLE
```
Comment (sushain97) [inline config.conf.example]: Why was this changed? The previous construction was clearer.

**Comment by sushain97** [comment]
@shardulc what do you think makes for a good file name? `constants.conf`?

File: assets/js/util.js [hunk 10]
```diff
@@ -7,16 +7,6 @@ var SPACE_KEY_CODE = 32, ENTER_KEY_CODE = 13,
     HTTP_OK_CODE = 200, HTTP_BAD_REQUEST_CODE = 400,
     XHR_LOADING = 3, XHR_DONE = 4;
 
-var TEXTAREA_AUTO_RESIZE_MINIMUM_WIDTH = 768,
-    BACK_TO_TOP_BUTTON_ACTIVATION_HEIGHT = 300,
-    APY_REQUEST_URL_THRESHOLD_LENGTH = 2000, // maintain 48 characters buffer for generated parameters
```
Comment (shardulc) [inline assets/js/util.js]: I don't know what the comment is about but you should probably preserve it in the config file.

File: tools/read-conf.py [hunk 11]
```diff
@@ -7,57 +7,105 @@
...
+
 
 def check_config(conf, result):
     # Some error checking:
     for section in conf.sections():
-        if section not in ['APY', 'REPLACEMENTS']:
+        if section not in ['APY', 'PERSISTENCE', 'REPLACEMENTS', 'TRANSLATOR', 'UTIL']:
             raise configparser.Error("\nUnknown section [%s]" % (section,))
 
+    # TODO: either remove or check for all sections
```
Comment (shardulc) [inline tools/read-conf.py]: Instead of checking the result after it's been generated, you could: 1. do `check_config` before you populate `results` and 2. check whether each item in all sections of `config` is present in `dtypes`.

Comment (shardulc) [reply tools/read-conf.py]: More helpful error messages throughout this code would be good, too.

File: tools/read-conf.py [hunk 12]
```diff
@@ -7,57 +7,105 @@
...
 
     return True
 
-def load_conf(filename):
-    conf = configparser.ConfigParser()
-    with open(filename, 'r') as f:
-        conf.read_file(f)
-    conf_APY = conf['APY']
-    result = {
-        'HTML_URL'                       : conf_APY.get('HTML_URL', fallback="http://www.apertium.org"),
```
Comment (shardulc) [inline tools/read-conf.py]: You forgot about the fallbacks this time :)

File: tools/read-conf.py [hunk 13]
```diff
@@ -84,6 +140,7 @@ def print_keyval(result, args):
 if __name__ == "__main__":
     parser = argparse.ArgumentParser(description='Load config, print stuff')
     parser.add_argument('-c', '--config', default='config.conf', help='Config file name (default: config.conf)')
+    parser.add_argument('-C', '--custom', default='custom.conf', help='Customization file name (default: custom.conf)')
```
Comment (shardulc) [inline tools/read-conf.py]: @sushain97 What about `internal-constants.conf`? (@Androbin don't forget to rename variables once we decide.)

File: config.conf.example (hunk 9 above)
Comment (Androbin) [reply config.conf.example]: At first, I moved the fallback into the custom format file. But when I turned it into a conf file, I set the initial values directly to the fallbacks. I will move the non-`None` fallbacks back into `read-conf.py`, a simple `dict` should do.

File: assets/js/util.js (hunk 10 above)
Comment (Androbin) [reply assets/js/util.js]: Sure.

File: tools/read-conf.py (hunk 11 above)
Comment (Androbin) [reply tools/read-conf.py]: I did not write either the test nor the error messages. But I will try to implement your suggestions.

File: tools/read-conf.py (hunk 12 above)
Comment (Androbin) [reply tools/read-conf.py]: I had them in the custom format file but when I turned it into a config file, I set the initial values to the fallbacks. I will add a `dict` for all non-`None` fallbacks in `read-conf.py`

File: tools/read-conf.py (hunk 13 above)
Comment (Androbin) [reply tools/read-conf.py]: But "internal" sounds like the path should not be configurable at all.

File: assets/js/util.js (hunk 10 above)
Comment (Androbin) [reply assets/js/util.js]: I think `2000 = 2^11 - 48` btw.

**Comment by Androbin** [comment]
@shardulc @sushain97 I hope this approach is of both your liking. I do now list and describe all constants in `read-conf.py` and can now iterate over this `dict` instead of iterating over the values in the `conf` files.

File: assets/js/util.js (hunk 10 above)
Comment (sushain97) [reply assets/js/util.js]: I don't actually think it should be configurable.

File: tools/read-conf.py (hunk 13 above)
Comment (sushain97) [reply tools/read-conf.py]: What's wrong with `constants.conf`? Seems OK to me.

File: tools/read-conf.py [hunk 14]
```diff
@@ -7,58 +7,115 @@
...
+    array = get_string_array(conf, key)
+    return fallback if array is None else [int(x) for x in array]
+
+
+# Constant descriptors
+constants = {
+    # section name
+    'APY': {
+        # CONSTANT_NAME: [parser]
+        # CONSTANT_NAME: [parser, fallback]
```
Comment (sushain97) [inline tools/read-conf.py]: This should not be an list, a tuple is a far far better choice.

File: assets/js/util.js (hunk 10 above)
Comment (Androbin) [reply assets/js/util.js]: Okay

File: tools/read-conf.py (hunk 13 above)
Comment (Androbin) [reply tools/read-conf.py]: Well, `config.conf` contains "constants" as well.

File: tools/read-conf.py (hunk 14 above)
Comment (Androbin) [reply tools/read-conf.py]: Sure

**Comment by sushain97** [comment]
@shardulc this looks mostly OK to me aside from the tuple/list comment above. I think if you can verify that it doesn't break any of the modes, its safe to approve the task. We can merge after we figure out naming of the second config file and take a last close look at the code.

File: tools/read-conf.py (hunk 13 above)
Comment (sushain97) [reply tools/read-conf.py]: Fair enough. Hmm... `advanced-config.conf`?

**Comment by Androbin** [comment]
@sushain97 I can't quite figure out how to rewrite it using tuples.
Python's unpacking rules are somewhat mean to me.

**Comment by sushain97** [comment]
What's the error? You need to use (firstitem, ) to make a tuple with one
item.

**Comment by Androbin** [comment]
> You need to use `(firstitem, )` to make a tuple with one
Oh, didn't know that. Thanks, will try again.

File: tools/read-conf.py [hunk 15]
```diff
@@ -67,6 +124,12 @@ def load_conf(filename):
         'ANALYZERS': None,
         'TAGGERS': None
     }
+
+    for section_name, section_desc in constants.items():
+        for constant_name, constant_desc in section_desc.items():
+            dtype, fallback, *_ = (*constant_desc, None)
```
[the remark this answers is not in the document]
Comment (shardulc) [reply tools/read-conf.py]: Um why not just `dtype, fallback = constant_desc`?

[the remark this answers is not in the document]
Comment (Androbin) [reply tools/read-conf.py]: Does not work when no fallback is specified.

[the remark this answers is not in the document]
Comment (Androbin) [reply tools/read-conf.py]: But I got an idea. I will add a function to pad the tupel with `None`.

**Comment by sushain97** [comment]
This just screams hack. I can't think of an elegant way to destructure both
a pair and singleton simultaneously in Python so I suggest just going with
the more verbose conditional option.

**Comment by Androbin** [comment]
@sushain97 What you think about my padding approach?

**Comment by shardulc** [comment]
@sushain97 @Androbin I think I have an elegant solution. What if:
 - every single constant in `read-conf.py` had a fallback whose value is the default for html-tools (i.e. the values that we know to work)
 - every single line in the file `advanced-config.conf` is commented, and some text explains how you would customize the constants by uncommenting and changing them

**Comment by Androbin** [comment]
@shardulc Do you propose to dissolve `config.conf.example` into the fallbacks in `read-conf.py` ?

**Comment by sushain97** [comment]
I actually don't like that approach too much because it requires editing a
Python file to change the interface. I think it makes everything a little
too connected and feels very magical.

I think the verbose method of figuring out whether there's a fallback or
not is fine. It's like 2 lines of code extra.

**Comment by shardulc** [comment]
> requires editing a Python file to change the interface
No, I edited my comment to make it more clear.

**Comment by Androbin** [comment]
@shardulc What is keeping us from combining `config,conf` and `custom.conf` anymore then?

**Comment by sushain97** [comment]
> No, I edited my comment to make it more clear.

A developer would have to modify the Python to change how the interface looks by default. This seems not great to me with regards to separation of concerns.

**Comment by Androbin** [comment]
@sushain97 As far as I understand @shardulc we'd essentially dissolve `config.conf.example` into the fallbacks in `read-conf.py`

**Comment by sushain97** [comment]
Well, I think he means dissolving `advanced-config.conf` into `read-conf.py`. We can't remove `config.conf.example`, it would seriously break backwards compatibility.

**Comment by shardulc** [comment]
> dissolving `advanced-config.conf` into `read-conf.py`

Yes, that's what I meant.

> A developer would have to modify the Python

I was thinking only goavki would modify the Python when required to tweak constants. If downstream users want to tweak constants, they'd be recommended to create/change `advanced-config.conf` instead so that they can keep their tweaks even when we push different fallbacks. (Changing the conf file does not require a developer either.)

**Comment by sushain97** [comment]
> I was thinking only goavki would modify the Python when required to tweak constants.

Right. This is my issue because we lose separation of concerns. Right now the Python is just a tool without any semantic meaning (it has fallbacks but those are duplicated in `config.conf.example`). If we implemented this idea, this would no longer be the case. We could store the defaults in the JS and hydrate the configuration with those defaults when they're missing, i.e. `advanced-config.conf.example` would have commented out constants and it could be moved to `advanced-config.conf`.

Either way, I think we should approve @Androbin's task and figure this out amongst ourselves. We should have asked for a concrete plan as one task and implementation as another since there are lots of ways to do this :)

**Comment by shardulc** [comment]
Ah, OK. @Androbin I'm approving your task then.

**Comment by Androbin** [comment]
Any updates on this yet?

**Comment by shardulc** [comment]
@Androbin No 'updates', but discussion is [over here](https://github.com/goavki/apertium-html-tools/issues/158#issuecomment-349368088). If you want to resolve those comments, you can do it as part of another task. (It won't be enough for one task because in my opinion it's only a few changes, you can pair it with another easy issue.)

**Comment by sushain97** [comment]
I haven't had a chance to look at it carefully yet. As part of the flow
updates, the approach to this will have to change a bit so let's not make
any changes yet.

**Comment by Androbin** [comment]
@sushain97 When the `flow` changes are done, you can give an outline of these desired changes.

**Comment by Androbin** [comment]
> As part of the flow updates, the approach to this will have to change a bit so let's not make any changes yet.

@sushain97 Could you give an outline of these desired changes.
